### PR TITLE
Subscription dance: backoff, logging, and test coverage

### PR DESCRIPTION
**Pull request by pdyraga**
Refs https://github.com/keep-network/keep-ecdsa/issues/680

This PR is the first one in the series of PRs refactoring Ethereum event subscriptions mechanism I plan to open in the next few days.

You can see it in action here: https://github.com/keep-network/keep-ecdsa/pull/663.

So far resubscription logging was implemented in contract binding templates. There were three problems with that mechanism:
1. No resubscribe backoff. Each failed subscription was retried 5 seconds after it failed. For many individual subscriptions ECDSA client is opening, a massive retry with no backoff could be interpreted as misbehavior by a third party Ethereum provider and further attempts could be completely blocked. Some operators experienced it with Alchemy.
2. All the logging was on the warning level. With the lack of backoff mentioned in the previous point, this could produce a deadly mixture - keep client could be trying to reconnect for a long time, Ethereum client could be rejecting those attempts interpreting them as DoS/misbehavior, and operator received only warnings with no single error.
3. Lack of test coverage for the resubscription mechanism. All the subscription code was placed in templates used to generate Go contract bindings and was very hard - if possible at all - to test.

Here we address all those problems.

Instead of implementing resubscriptions on our side, we wrap the code from `github.com/ethereum/go-ethereum/event` with some additional logging logic. This code is used from contract templates to keep the subscription alive. We lean on `go-ethereum` to do those resubscriptions right and we cover our logic addition in unit tests on our side.

The code from `go-ethereum` implements backoffs, increasing the delay twice until it reaches the maximum backoff time which is set in our bindings to 2 minutes.

The new code in `ethutil` wrapping `go-ethereum`'s resubscriber, allows to log an error if the subscription is dropped too often which may indicate problems with Ethereum client. The threshold is set to 15 minutes.

### CODE REVIEW

**Review by nkuba — COMMENTED** [review]
First part of comments.

**Review by lukasz-zimnoch — CHANGES_REQUESTED** [review]

File: pkg/chain/ethereum/ethutil/resubscribe_test.go [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+		alertThreshold,
+		func(elapsed time.Duration) { thresholdViolated <- elapsed },
+		func(err error) { subscriptionFailed <- err },
+	)
+	<-subscription.Err()
+
+	// Subscription failed one time so there should be one error in the channel.
+	subscriptionFailCount := len(subscriptionFailed)
+	if subscriptionFailCount != 1 {
+		t.Fatalf(
```
Comment (nkuba) [inline pkg/chain/ethereum/ethutil/resubscribe_test.go:42]: Can we replace `t.Fatalf` with `t.Errorf` in all added tests so the test reports all verification failures instead of aborting on the first one?

File: pkg/chain/ethereum/ethutil/resubscribe_test.go [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+	// blocked by the lack of channel receivers on non-buffered channels.
+	expectedResubscriptionCalls := plannedSubscriptionFailures + 1
+	if resubscribeFnCalls != expectedResubscriptionCalls {
+		t.Fatalf(
+			"resubscription called [%v] times, expected [%v]",
+			resubscribeFnCalls,
+			expectedResubscriptionCalls,
+		)
+	}
+}
```
Comment (nkuba) [inline pkg/chain/ethereum/ethutil/resubscribe_test.go:280]: Thoughts on adding a test where we check that if the subscription is cancelled callback functions won't be called anymore?

File: tools/generators/ethereum/contract_events.go.tmpl [hunk 3]
```diff
@@ -37,119 +37,61 @@ func ({{$contract.ShortVar}} *{{$contract.Class}}) Past{{$event.CapsName}}Events
...
 
-	return subscription.NewEventSubscription(unsubscribeCallback), nil
+	sub := ethutil.WithResubscription(
+		{{$contract.ShortVar}}SubscriptionBackoffMax,
+		subscribeFn,
+		{{$contract.ShortVar}}SubscriptionAlertThreshold,
+		func(elapsed time.Duration) {
+			{{$logger}}.Errorf(
+					"subscription to event {{$event.CapsName}} had to be "+
+						"retried [%v] since the last attempt; please inspect "+
```
Comment (nkuba) [inline tools/generators/ethereum/contract_events.go.tmpl:76]: Will it print fine with `%v` shouldn't it be `%s`?
```suggestion
						"retried [%s] since the last attempt; please inspect "+
```

File: tools/generators/ethereum/contract_events.go.tmpl [hunk 4]
```diff
@@ -37,119 +37,61 @@ func ({{$contract.ShortVar}} *{{$contract.Class}}) Past{{$event.CapsName}}Events
...
+						"Ethereum client connectivity",
```
Comment (nkuba) [inline tools/generators/ethereum/contract_events.go.tmpl:77]: Maybe we should drop `client`? Someone can connect to an ethereum client but others can use providers like Infura or Alchemy.

```suggestion
						"Ethereum connectivity",
```

File: tools/generators/ethereum/contract_events.go.tmpl [hunk 5]
```diff
@@ -37,119 +37,61 @@ func ({{$contract.ShortVar}} *{{$contract.Class}}) Past{{$event.CapsName}}Events
...
+		)
 	}
 
-	return subscription.NewEventSubscription(unsubscribeCallback), nil
+	sub := ethutil.WithResubscription(
+		{{$contract.ShortVar}}SubscriptionBackoffMax,
+		subscribeFn,
+		{{$contract.ShortVar}}SubscriptionAlertThreshold,
+		func(elapsed time.Duration) {
+			{{$logger}}.Errorf(
```
Comment (nkuba) [inline tools/generators/ethereum/contract_events.go.tmpl:74]: Error or Warn, that is the question.

File: tools/generators/ethereum/contract_events.go.tmpl [hunk 6]
```diff
@@ -37,119 +37,61 @@ func ({{$contract.ShortVar}} *{{$contract.Class}}) Past{{$event.CapsName}}Events
...
+					"subscription to event {{$event.CapsName}} had to be "+
+						"retried [%v] since the last attempt; please inspect "+
+						"Ethereum client connectivity",
+					elapsed,
+				)
+		},
+		func(err error) {
+			{{$logger}}.Errorf(
```
Comment (nkuba) [inline tools/generators/ethereum/contract_events.go.tmpl:82]: This vs L74. One of them could be Error and the other Warn. Maybe this could be a Warn and the other one Error?

File: pkg/chain/ethereum/ethutil/resubscribe_test.go [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+	thresholdViolated := make(chan time.Duration, 10)
+	subscriptionFailed := make(chan error, 10)
+	subscription := WithResubscription(
+		backoffMax,
+		resubscribeFn,
+		alertThreshold,
+		func(elapsed time.Duration) { thresholdViolated <- elapsed },
+		func(err error) { subscriptionFailed <- err },
+	)
+	<-subscription.Err()
```
Comment (nkuba) [inline pkg/chain/ethereum/ethutil/resubscribe_test.go:37]: How does it work? When does it resolve? Seems like resubscription is kept forever but when eventually it returns an error?

File: pkg/chain/ethereum/ethutil/resubscribe.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+// thresholdViolatedFn and subscriptionFailedFn calls are executed in a separate
+// goroutine and thus are non-blocking.
+func WithResubscription(
+	backoffMax time.Duration,
+	resubscribeFn event.ResubscribeFunc,
+	alertThreshold time.Duration,
+	thresholdViolatedFn func(time.Duration),
+	subscriptionFailedFn func(error),
+) event.Subscription {
+	lastAttempt := time.Time{}
```
[the remark this answers is not in the document]
Comment (nkuba) [reply pkg/chain/ethereum/ethutil/resubscribe.go:37]: Yeah, seems like my comment doesn't make sense. Sorry about that!

File: tools/generators/ethereum/contract.go.tmpl [hunk 9]
```diff
@@ -25,6 +25,21 @@ import (
...
+
+    // Threshold below which event resubscription emits an error to the logs.
+    // WS connection can be dropped at any moment and event resubscription will
+    // follow. However, if WS connection for event subscription is getting
+    // dropped too often, it may indicate something is wrong with Ethereum
+    // client. This constant defines the minimum lifetime of an event
+    // subscription required before the subscription failure happens and
+    // resubscription follows so that the resubscription does not emit an error
+    // to the logs alerting about potential problems with Ethereum client.
+    {{.ShortVar}}SubscriptionAlertThreshold = 15 * time.Minute
```
Comment (Shadowfiend) [inline tools/generators/ethereum/contract.go.tmpl:40]: 🤔 should this be per-contract?

File: pkg/chain/ethereum/ethutil/resubscribe.go [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+	wrappedResubscribeFn := func(ctx context.Context) (event.Subscription, error) {
+		now := time.Now()
+		elapsed := now.Sub(lastAttempt)
+		if elapsed < alertThreshold {
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply pkg/chain/ethereum/ethutil/resubscribe.go:41]: There is no `Before` function defined for `time.Duration`. I think it's fine to use `<` in this case because both `elapsed` and `alertThreshold` are `time.Duration`.

File: pkg/chain/ethereum/ethutil/resubscribe.go [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+// subscriptionFailedFn is called with the underlying error.
+//
+// thresholdViolatedFn and subscriptionFailedFn calls are executed in a separate
+// goroutine and thus are non-blocking.
+func WithResubscription(
+	backoffMax time.Duration,
+	resubscribeFn event.ResubscribeFunc,
+	alertThreshold time.Duration,
+	thresholdViolatedFn func(time.Duration),
+	subscriptionFailedFn func(error),
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply pkg/chain/ethereum/ethutil/resubscribe.go:35]: How important is this change for you? I started with something similar but then switched to shorter names to keep this code easier to read.

[the remark this answers is not in the document]
Comment (pdyraga) [reply pkg/chain/ethereum/ethutil/resubscribe.go:35]: This function is also called if the first-ever subscription attempt fails. My reasoning was that the first attempt is *a subscription*  and even if we have to resubscribe later, resubscription attempts are also... *subscription* attempts, just repeated. That's why I named it `subscriptionFailedFn`.

In fact `resubscribeFn` could be renamed to `subscribeFn` but given that the function type from `github.com/ethereum/go-ethereum/event` is `event.ResubscribeFn` I left it as `resubscribeFn`. Happy to change it though if you think it makes sense.

File: pkg/chain/ethereum/ethutil/resubscribe_test.go [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+			return nil, fmt.Errorf("wherever I go, he goes")
+		}
+		delegate := event.NewSubscription(func(unsubscribed <-chan struct{}) error {
+			return nil
+		})
+		return delegate, nil
+	}
+
+	thresholdViolated := make(chan time.Duration, 10)
+	subscriptionFailed := make(chan error, 10)
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply pkg/chain/ethereum/ethutil/resubscribe_test.go:29]: Added:
```
	// Using buffered channels to do not block writes.
	// There should never be a need to write more to those channels if the code
	// under the test works as expected.
```

File: tools/generators/ethereum/contract_events.go.tmpl [hunk 13]
```diff
@@ -37,119 +37,61 @@ func ({{$contract.ShortVar}} *{{$contract.Class}}) Past{{$event.CapsName}}Events
...
+					"subscription to event {{$event.CapsName}} failed "+
+						"with error: [%v]; resubscription attempt will be "+
+						"performed",
+					err,
+				)
+		},
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply tools/generators/ethereum/contract_events.go.tmpl]: See eb5ee3d0578ab0adde55c7e6d732c8d7f40f5e83 please.

File: pkg/chain/ethereum/ethutil/resubscribe_test.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (pdyraga) [reply pkg/chain/ethereum/ethutil/resubscribe_test.go:32]: See 44412b59cf176977a2f301cf75ed5d57380846ad please.

File: pkg/chain/ethereum/ethutil/resubscribe_test.go [hunk 14]
```diff
@@ -0,0 +1,280 @@
...
+			return nil, fmt.Errorf("this is the way")
+		}
+		delegate := event.NewSubscription(func(unsubscribed <-chan struct{}) error {
+			return nil
+		})
+		return delegate, nil
+	}
+
+	thresholdViolated := make(chan time.Duration, 10)
+	subscriptionFailed := make(chan error, 10)
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply pkg/chain/ethereum/ethutil/resubscribe_test.go:87]: See 44412b59cf176977a2f301cf75ed5d57380846ad please.

File: pkg/chain/ethereum/ethutil/resubscribe_test.go [hunk 15]
```diff
@@ -0,0 +1,280 @@
...
+			return nil, fmt.Errorf("i have spoken")
+		}
+		delegate := event.NewSubscription(func(unsubscribed <-chan struct{}) error {
+			return nil
+		})
+		return delegate, nil
+	}
+
+	thresholdViolated := make(chan time.Duration, 10)
+	subscriptionFailed := make(chan error, 10)
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply pkg/chain/ethereum/ethutil/resubscribe_test.go:156]: See 44412b59cf176977a2f301cf75ed5d57380846ad please.

File: pkg/chain/ethereum/ethutil/resubscribe.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (pdyraga) [reply pkg/chain/ethereum/ethutil/resubscribe.go:35]: This was also brought up by @nkuba here: https://github.com/keep-network/keep-common/pull/62#discussion_r558288557. I will do `resubscribeFn` -> `subscribeFn` rename.

File: pkg/chain/ethereum/ethutil/resubscribe.go [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+// subscription.
+//
+// In case of an error returned by the wrapped subscription function,
+// subscriptionFailedFn is called with the underlying error.
+//
+// thresholdViolatedFn and subscriptionFailedFn calls are executed in a separate
+// goroutine and thus are non-blocking.
+func WithResubscription(
+	backoffMax time.Duration,
+	resubscribeFn event.ResubscribeFunc,
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply pkg/chain/ethereum/ethutil/resubscribe.go:32]: Yes, it makes sense. We discussed it earlier here: https://github.com/keep-network/keep-common/pull/62#discussion_r565173008

tl;dr; I left it as `resubscribeFn` because of the type name but was not particularly happy about it. Since you are bringing this up, I am happy to do the rename.

File: pkg/chain/ethereum/ethutil/resubscribe.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (pdyraga) [reply pkg/chain/ethereum/ethutil/resubscribe.go:35]: See e81b7de053718e5b4d6c67275f7528529af92854 please.

File: pkg/chain/ethereum/ethutil/resubscribe.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (pdyraga) [reply pkg/chain/ethereum/ethutil/resubscribe.go]: See e81b7de053718e5b4d6c67275f7528529af92854 please.

File: tools/generators/ethereum/contract_events.go.tmpl [hunk 17]
```diff
@@ -37,119 +37,61 @@ func ({{$contract.ShortVar}} *{{$contract.Class}}) Past{{$event.CapsName}}Events
 
 func ({{$contract.ShortVar}} *{{$contract.Class}}) Watch{{$event.CapsName}}(
 	success {{$contract.FullVar}}{{$event.CapsName}}Func,
-	fail func(err error) error,
 	{{$event.IndexedFilterDeclarations -}}
-) (subscription.EventSubscription, error) {
-    errorChan := make(chan error)
-    unsubscribeChan := make(chan struct{})
+) (subscription.EventSubscription) {
+	eventOccurred := make(chan *abi.{{$contract.AbiClass}}{{$event.CapsName}})
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply tools/generators/ethereum/contract_events.go.tmpl:42]: Agreed though this line will go away in #63 so I suggest we leave it as is here.

File: tools/generators/ethereum/contract_events.go.tmpl [hunk 18]
```diff
@@ -37,119 +37,61 @@ func ({{$contract.ShortVar}} *{{$contract.Class}}) Past{{$event.CapsName}}Events
...
-	if err != nil {
-		close(eventChan)
-		return eventSubscription, fmt.Errorf(
-			"error creating watch for {{$event.CapsName}} events: [%v]",
-			err,
-		)
-	}
-
-	var subscriptionMutex = &sync.Mutex{}
+	ctx, cancel := context.WithCancel(context.Background())
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply tools/generators/ethereum/contract_events.go.tmpl]: Please see 7a5fcf12f04fefbab6137b71a21e483050482e70.

File: tools/generators/ethereum/contract_events.go.tmpl (hunk 5 above)
Comment (pdyraga) [reply tools/generators/ethereum/contract_events.go.tmpl]: Both of them should be logged on the error level. `thresholdViolatedFn` is called when resubscribe attempt had to be performed too soon. If the connection is getting broken every few minutes, it is clear something is wrong and there is a chance the client will not be able to receive events on time. This should be escalated.

File: tools/generators/ethereum/contract_events.go.tmpl (hunk 6 above)
Comment (pdyraga) [reply tools/generators/ethereum/contract_events.go.tmpl]: Both of them should be logged on the error level. `subscriptionFailedFn` is called when a subscribe attempt fails right away. It is fine if ws connection gets broken after several days and the subscription has to be retried. But it is very alarming if it fails right after the connect attempt.

File: pkg/chain/ethereum/ethutil/resubscribe_test.go [hunk 19]
```diff
@@ -0,0 +1,280 @@
...
+	subscriptionFailCount := len(subscriptionFailed)
+	if subscriptionFailCount != 1 {
+		t.Fatalf(
+			"subscription failure reported [%v] times, expected [1]",
+			subscriptionFailCount,
+		)
+	}
+
+	// That failure should refer the original error.
+	expectedFailMessage := "wherever I go, he goes"
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply pkg/chain/ethereum/ethutil/resubscribe_test.go]: Sure. See eb2cc4af228eca465b33f4ad92ce6c1fb295368d please.

File: pkg/chain/ethereum/ethutil/resubscribe_test.go (hunk 1 above)
Comment (pdyraga) [reply pkg/chain/ethereum/ethutil/resubscribe_test.go:46]: The reason I used `Fatalf` in this case is that we should not execute the other check if we do not want to risk the test to hang. If this channel is empty, `err := <-subscriptionFailed` execution in the next line will hang. For other cases, I do agree we can use `Errorf`.

Comment (pdyraga) [reply pkg/chain/ethereum/ethutil/resubscribe_test.go:46]: See 4841e64672d7215b060ca21758d9576d44f292e5 please.

File: pkg/chain/ethereum/ethutil/resubscribe_test.go [hunk 20]
```diff
@@ -0,0 +1,280 @@
...
+	)
+	<-subscription.Err()
+
+	// Subscription failed one time so there should be one error in the channel.
+	subscriptionFailCount := len(subscriptionFailed)
+	if subscriptionFailCount != 1 {
+		t.Fatalf(
+			"subscription failure reported [%v] times, expected [1]",
+			subscriptionFailCount,
+		)
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply pkg/chain/ethereum/ethutil/resubscribe_test.go:49]: I am not sure about it, same for the other mentioned cases. For simple assertions like this one, we can fit everything in one line as there is no risk that expected/actual values comparison will be hard. And the `Errof`/`Fatalf` string looks much clearer in the code.

I would suggest that we do a single-line assertion failed message in such cases and we split to multiple lines when we e.g. compare error messages or big integers or hex byte arrays. WDYT?

File: tools/generators/ethereum/contract.go.tmpl (hunk 9 above)
Comment (pdyraga) [reply tools/generators/ethereum/contract.go.tmpl:40]: I think it should not be but I do not have a good place to put it at. :grimacing: Maybe `subscribe_opts.go` from `ethutil` introduced in the next PR #63 is a good candidate?

File: pkg/chain/ethereum/ethutil/resubscribe.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply pkg/chain/ethereum/ethutil/resubscribe.go:35]: Sure, I'm fine with `subscribeFn`

File: pkg/chain/ethereum/ethutil/resubscribe_test.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply pkg/chain/ethereum/ethutil/resubscribe_test.go:49]: Sure, good idea!

**Review by lukasz-zimnoch — COMMENTED** [review]
No more comments! Leaving the rest for @nkuba

File: tools/generators/ethereum/contract_events.go.tmpl (hunk 5 above)
Comment (nkuba) [reply tools/generators/ethereum/contract_events.go.tmpl]: OK, so we won't log an error every time resubscription happens but only if it happens too soon?

File: pkg/chain/ethereum/ethutil/resubscribe_test.go (hunk 7 above)
Comment (pdyraga) [reply pkg/chain/ethereum/ethutil/resubscribe_test.go:41]: It does resolve here:

https://github.com/ethereum/go-ethereum/blob/2e5d14170846ae72adc47467a1129e41d6800349/event/subscription.go#L52

The function passed eventually to `event.NewSubscription` by our test subscribe function is non-blocking, there is nothing to produce and that `defer` will be called closing channel returned by `Err()`. Similar pattern is used in `go-ethereum` tests:

https://github.com/ethereum/go-ethereum/blob/2e5d14170846ae72adc47467a1129e41d6800349/event/subscription_test.go#L98

File: tools/generators/ethereum/contract_events.go.tmpl (hunk 5 above)
Comment (pdyraga) [reply tools/generators/ethereum/contract_events.go.tmpl]: Correct. We log error in two cases:
- resubscription had to happen too soon (threshold violated),
- subscription (resubscription) failed right after it's been requested.

Nothing is logged when the subscription fails after e.g. ten days. This is how `go-ethereum/event` `Resubscribe` approaches it. However, we modified it in https://github.com/ethereum/go-ethereum/pull/22191 and I suggest we log a warning. This is out of the scope of this PR and will require `go-ethereum` dependency bump.

File: tools/generators/ethereum/contract_events.go.tmpl (hunk 4 above)
Comment (pdyraga) [reply tools/generators/ethereum/contract_events.go.tmpl]: Sure, please see c777eab88a4a02e21b166c54e85a48055f2a74f9.

File: tools/generators/ethereum/contract_events.go.tmpl (hunk 3 above)
Comment (pdyraga) [reply tools/generators/ethereum/contract_events.go.tmpl]: It will work fine with both, I can replace `%v` with `%s` for clarity.

Comment (pdyraga) [reply tools/generators/ethereum/contract_events.go.tmpl]: See 64cab832ccbd7ae02b6e87da48ff65bf92fd8c1f please.

File: pkg/chain/ethereum/ethutil/resubscribe_test.go (hunk 2 above)
Comment (pdyraga) [reply pkg/chain/ethereum/ethutil/resubscribe_test.go:289]: Do you think it is needed? Those functions are called from the inside of a wrapped subscribe function and I expect `go-ethereum` `Resubscribe` to be implemented right and do not call subscribe again if the given subscription was cancelled.

Even if we treat this code as a black box - which I usually prefer to do - I am not sure if it is possible to write such a test - I will unsubscribe and I can't get subscribe function to be called again, it does not make sense 🤔

**Comment by pdyraga** [comment]
I updated https://github.com/keep-network/keep-ecdsa/pull/663 to point to the version with all review comments addressed, which is c6adefdc082b4ee80aa6da0075e0f4baa8531f70 - the last commit.

**Review by lukasz-zimnoch — APPROVED** [review]

**Review by nkuba — APPROVED** [review]

File: tools/generators/ethereum/contract.go.tmpl (hunk 9 above)
Comment (nkuba) [reply tools/generators/ethereum/contract.go.tmpl:40]: I'm going to merge this PR. Please let's handle this change in #63.
